tests/zkevm/test_worst_stateful_opcodes.py

```diff
@@ -475,3 +479,230 @@ def test_worst_extcodecopy_warm(
+        deployed_contract_addresses.append(deployed_contract_address)
+
+    blockchain_test(
+        genesis_environment=env,
+        pre=pre,
+        post=post,
+        blocks=[
+            Block(txs=[contracts_deployment_tx]),
+            Block(txs=[opcode_tx]),
+        ],
+        exclude_full_post_state_in_output=True,
+    )
+
+
+@pytest.mark.valid_from("Cancun")
+@pytest.mark.parametrize("value_bearing", [True, False])
+def test_worst_selfdestruct_created(
+    state_test: StateTestFiller,
+    pre: Alloc,
+    value_bearing: bool,
+):
+    """
+    Test running a block with as many SELFDESTRUCTs as possible for deployed contracts in
+    the same transaction.
+    """
+    env = Environment()
+    pre.fund_address(env.fee_recipient, 1)
+
+    # SELFDESTRUCT(COINBASE) contract deployment
+    initcode = Op.MSTORE8(0, 0x41) + Op.MSTORE8(1, 0xFF) + Op.RETURN(0, 2)
+    code = (
+        Op.MSTORE(0, initcode.hex())
+        + While(
+            body=Op.POP(
+                Op.CALL(
+                    address=Op.CREATE(
+                        value=1 if value_bearing else 0,
+                        offset=32 - len(initcode),
+                        size=len(initcode),
+                    )
+                )
+            ),
+            # Stop before we run out of gas for the whole tx execution.
+            # The value was found by trial-error rounded to the next 1000 multiple.
+            condition=Op.GT(Op.GAS, 10_000),
+        )
+        + Op.SSTORE(0, 42)  # Done for successful tx execution assertion below.
+    )
+    # The 0 storage slot is initialize to avoid creation costs in SSTORE above.
+    code_addr = pre.deploy_contract(code=code, balance=100_000, storage={0: 1})
+    code_tx = Transaction(
+        to=code_addr,
+        gas_limit=env.gas_limit,
+        gas_price=10,
+        sender=pre.fund_eoa(),
+    )
+
+    post = {code_addr: Account(storage={0: 42})}  # Check for successful execution.
+    state_test(
+        env=env,
+        pre=pre,
+        post=post,
+        tx=code_tx,
+    )
+
+
+@pytest.mark.valid_from("Cancun")
+@pytest.mark.parametrize("value_bearing", [True, False])
+def test_worst_selfdestruct_initcode(
+    state_test: StateTestFiller,
+    pre: Alloc,
+    value_bearing: bool,
+):
+    """Test running a block with as many SELFDESTRUCTs as possible executed in initcode."""
+    env = Environment()
+    pre.fund_address(env.fee_recipient, 1)
+
+    initcode = Op.SELFDESTRUCT(Op.COINBASE)
+    code = (
+        Op.MSTORE(0, initcode.hex())
+        + While(
+            body=Op.POP(
+                Op.CREATE(
+                    value=1 if value_bearing else 0,
+                    offset=32 - len(initcode),
+                    size=len(initcode),
+                )
+            ),
+            # Stop before we run out of gas for the whole tx execution.
+            # The value was found by trial-error rounded to the next 1000 multiple.
+            condition=Op.GT(Op.GAS, 12_000),
```

Comment:
This constant 12_000 has to be changed in the last commit which can feel surprising.
The reason is that the number should guarantee an extra iteration + the after-loop code.

If previously a 10_000 number worked, can be a red-herring since probably the iteration had say 15_000 of gas which is >10_000, but doing a code change can reveal this case.

This is why this strategy never convinced me. Probably a better way is to have a conditional that does not account for an extra loop and only the cost of logic after the loop.

Something to meditate.

Reply:
Maybe we can have another While-like construction that injects the GAS before and after assuming constant gas usage as we chatted before. But that's also more gas used in tracking stuff -- but maybe in some cases is fine.